crypto/cms/cms_ec.c

```diff
@@ -0,0 +1,411 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/cms.h>
+#include <openssl/err.h>
+#include <openssl/decoder.h>
+#include "cms_local.h"
+#include "crypto/evp.h"
+
+static EVP_PKEY *pkey_type2param(int ptype, const void *pval,
+                                 OPENSSL_CTX *libctx, const char *propq)
+{
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *pctx = NULL;
+
+    if (ptype == V_ASN1_SEQUENCE) {
+        const ASN1_STRING *pstr = pval;
+        const unsigned char *pm = pstr->data;
+        int pmlen = pstr->length;
+        OSSL_DECODER_CTX *ctx = NULL;
+        BIO *membio = NULL;
+
+        /* TODO(3.0): Need to be able to specify here that only params will do */
+        ctx = OSSL_DECODER_CTX_new_by_EVP_PKEY(&pkey, "DER", "EC", libctx,
+                                               propq);
+        membio = BIO_new_mem_buf(pm, pmlen);
+        OSSL_DECODER_from_bio(ctx, membio);
+        BIO_free(membio);
+        OSSL_DECODER_CTX_free(ctx);
+    } else if (ptype == V_ASN1_OBJECT) {
+        const ASN1_OBJECT *poid = pval;
+        const char *groupname;
+
+        /*
+         * type == V_ASN1_OBJECT => the parameters are given by an asn1 OID
+         */
+
+        pctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", propq);
+
+        if (pctx == NULL)
+            goto err;
+        if (EVP_PKEY_paramgen_init(pctx) <= 0)
+            goto err;
+        groupname = OBJ_nid2sn(OBJ_obj2nid(poid));
+        if (groupname == NULL
+                || !EVP_PKEY_CTX_set_group_name(pctx, groupname)) {
+            CMSerr(0, CMS_R_DECODE_ERROR);
+            goto err;
+        }
+        if (EVP_PKEY_paramgen(pctx, &pkey) <= 0) {
+            EVP_PKEY_free(pkey);
+            pkey = NULL;
+        }
+    } else {
+        CMSerr(0, CMS_R_DECODE_ERROR);
+        goto err;
+    }
+
+    return pkey;
+
+ err:
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_CTX_free(pctx);
+    return NULL;
+}
+
+static int ecdh_cms_set_peerkey(EVP_PKEY_CTX *pctx,
+                                X509_ALGOR *alg, ASN1_BIT_STRING *pubkey)
+{
+    const ASN1_OBJECT *aoid;
+    int atype;
+    const void *aval;
+    int rv = 0;
+    EVP_PKEY *pkpeer = NULL;
+    const unsigned char *p;
+    int plen;
+
+    X509_ALGOR_get0(&aoid, &atype, &aval, alg);
+    if (OBJ_obj2nid(aoid) != NID_X9_62_id_ecPublicKey)
+        goto err;
+    /* If absent parameters get group from main key */
+    if (atype == V_ASN1_UNDEF || atype == V_ASN1_NULL) {
+        EVP_PKEY *pk;
+        pk = EVP_PKEY_CTX_get0_pkey(pctx);
+        if (pk == NULL)
+            goto err;
+
+        pkpeer = EVP_PKEY_new();
+        if (pkpeer == NULL)
+            goto err;
+        if (!EVP_PKEY_copy_parameters(pkpeer, pk))
+            goto err;
+    } else {
+        /* TODO(3.0): Should the get0_libctx/propq calls actually be public API? */
```

Comment:
I suspect the time for using TODO(3.0) is nearly over...

Reply:
I can remove it - but what do we think? Should these function calls be public API or not? I'm just thinking that if I needed it for CMS, then perhaps others will too.

Reply:
I think it is ok.. @levitte might have a different opinion :)

Reply:
I'm perfectly ok with it.  I made them internal because I've had quite a few pushbacks of the "do we really need to make this public?" type...

Reply:
> I made them internal because

I think you'll find *I* made them internal - they are new for this PR :-)

Anyway, I have now made them public API.